### PR TITLE
Test suite websocket

### PR DESCRIPTION
**Pull request by ricardopereira**
I was using `client.transition` to simulate _incomings of errors & connection state change_ and it is wrong because some of the logic is missed.

To change the connection states:
 - `client.onConnected(message: ARTProtocolMessage)`
 - `client.onDisconnected()`
 - `client.onSuspended()`
 - `client.onError(AblyTests.simulateError())`

With this PR we have:
 - `simulateIncomingNormalClose`
 - `simulateIncomingAbruptlyClose`
 - `simulateIncomingError`
 - `simulateLostConnection`

### CODE REVIEW

**Comment by ricardopereira** [comment]
@tcard What do you think?
@mattheworiordan I really needed this. There were several solutions around the test suite.

File: ably-ios/ARTRealtime.m [hunk 1]
```diff
@@ -446,7 +446,7 @@ -(void) cancelPingTimer {
     self.pingTimeout = nil;
 }
 
-- (void)onHeartbeat:(ARTProtocolMessage *)message {
+- (void)onHeartbeat {
```
Comment (mattheworiordan) [inline ably-ios/ARTRealtime.m]: Why did you remove the ProtocolMessage?

File: ably-ios/ARTRealtime+Private.h [hunk 2]
```diff
@@ -25,21 +25,17 @@ ART_ASSUME_NONNULL_BEGIN
 @property (readonly, strong, nonatomic) NSMutableArray *stateSubscriptions;
 
 // Transport Events
-- (void)onHeartbeat:(ARTProtocolMessage *)message;
-- (void)onConnected:(ARTProtocolMessage *)message withErrorInfo:(art_nullable ARTErrorInfo *)errorInfo;
-- (void)onDisconnected:(ARTProtocolMessage *)message;
-- (void)onError:(ARTProtocolMessage *)message withErrorInfo:(art_nullable ARTErrorInfo *)errorInfo;
+- (void)onHeartbeat;
+- (void)onConnected:(ARTProtocolMessage *)message;
+- (void)onDisconnected;
```
Comment (mattheworiordan) [inline ably-ios/ARTRealtime+Private.h]: Are you not missing `onClosed:(ART ProtocolMessage *)message`?

File: ably-iosTests/ARTTestUtil.m [hunk 3]
```diff
@@ -280,4 +281,11 @@ + (void)testRealtimeV2:(XCTestCase *)testCase withDebug:(BOOL)debug callback:(AR
     [testCase waitForExpectationsWithTimeout:[ARTTestUtil timeout] handler:nil];
 }
 
++ (ARTProtocolMessage *)simulateError {
```
Comment (mattheworiordan) [inline ably-iosTests/ARTTestUtil.m]: Should this not be named something `newErrorProtocolMessage` as it's not actually simulating anything until it's used, it's just creating an object

File: ablySpec/TestUtilities.swift [hunk 4]
```diff
@@ -114,6 +115,13 @@ class AblyTests {
         }
         return options
     }
+
+    class func simulateError() -> ARTProtocolMessage {
```
Comment (mattheworiordan) [inline ablySpec/TestUtilities.swift]: As above, this is not simulating an error but instead creating an error object

**Comment by mattheworiordan** [comment]
Looks fine, few comments re: naming

File: ably-ios/ARTRealtime+Private.h (hunk 2 above)
Comment (ricardopereira) [inline ably-ios/ARTRealtime+Private.h]: `onClosed` doesn't exist right from the beginning. I'm assuming the tests are calling the `close()` method.

File: ably-ios/ARTRealtime.m (hunk 1 above)
Comment (ricardopereira) [inline ably-ios/ARTRealtime.m]: Because the method was not using it.

File: ably-iosTests/ARTTestUtil.m (hunk 3 above)
Comment (ricardopereira) [inline ably-iosTests/ARTTestUtil.m]: Yes, I agree.

**Comment by tcard** [comment]
LGTM
